035-ca-abstraction.md

```diff
@@ -0,0 +1,315 @@
+class IssuedCertificate {
+  private final private byte[] key;
+  private final X509Certificate issuedCertificate;
+  private final List<X509Certificate> intermediateCaCertificates;
+  private final X509Certificate rootCaCertificate;
+  // ... constructor, accessors, etc.
+}
+```
+
+### Abstract process
+
+#### Process for certificate issuance
+
+Adopting the `CertificateIssuer` interface fundamentally changes _how_ the CO issues certificates. 
+Because of the current assumptions, the CO generates cluster CA certificates (if they're missing or expired) within a reconciliation, before it knows whether any EE certificate issuance is needed in that reconciliation. Then if it needs to issue EE certificates it just uses the CA certificates that must exist by that point.
+
+The steps required by the `CertificateIssuer` contract are more complicated, and will be described over the following sections. 
+The term "issuing operator" will be used to refer to the operator that's issuing the certificate. This could be the User or Cluster Operator.
+
+Before describing the process let's introduce two state machines. First the "EE certificate state machine"
+
+```
+  NEW // a certificate is needed for the end entity, but doesn't exist
+   |
+   v
+  REQUESTED // a certificate has been requested
+   |
+   v
+  PENDING // a certificate has been issued, but the root CA is not trusted yet
+   |
+   v
+  IN_USE // the certificate is in use
+   |
+   v
+  NOT_NEEDED // a certificate for the end entity is not needed any more
+```
+
+Then the "CA certificate trust state machine":
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_NEW // trusted everywhere it needs to be, but no EE certificates yet issued
+   |
+   v
+  TRUSTED_USED // trusted everywhere, EE certificates issued
+   |
+   v
+  PHASE_OUT // trusted everywhere, but being phased out
+   |
+   v
+  ZOMBIE // still trusted by some component, but not relied on by any party.
+```
+
+Now let's describe the processes.
+
+1. The issuing operator needs to figure out which EE certificates are missing (or in their renewal period, or otherwise need replacing, e.g. via annotation). This need for a certificate might exist independently of the end entity. For example, if we're going to scale up the Kafka cluster then we need a certificate for the new broker before the new broker exists. **The issuing operator puts the end entity in the `NEW` state.**
+2. (**See step 1 in the diagram.**) The issuing operator requests a new certificate. **The issuing operator puts the end entity in the `REQUESTED` state.**
+3. (**See step 2 in the diagram.**) The issuing operator cannot assume that issuance will happen quickly (i.e. that `pollForRequestedCertificate()` returns a non-empty `Optional` immediately), so it needs to be able persist the token obtained from `requestCertificate()` in a `Secret`, _and possibly complete the current reconciliation_ (i.e. issuance can be very asynchronous). 
```

Comment:
I don't want to couple the operator's reconciliation loop to an external system. It's not so much of a problem when bootstrapping a cluster, but blocking reconciliation because we need to renew a certificate feels like a bad idea, since it prevents other actions on the cluster. So rather than polling `pollForRequestedCertificate` within a single reconciliation, waiting for a non-empty result, the idea is to allow the reconciliation to proceed and call `pollForRequestedCertificate` again in the later reconciliation. That way we don't have to reason about how quickly a given certificate management system might issue a certificate. Its availability wouldn't matter much. There could be a human in the loop checking each CSR and the operator would still function just fine. It would be possible to implement the `external` issuer mentioned in the rejected alternatives and the operator would still function correctly.

Reply:
With CC there's nothing else that might need to be done by that operator, so it's OK to poll the CC REST indefinitely.  That's not the case with the Kafka operator.